### PR TITLE
Issue 327 - Support for List/Set Cardinality in Solr.

### PR DESCRIPTION
**Pull request by holleyism**
Signed-off-by: Adam Holley <holleya@us.ibm.com>

### CODE REVIEW

File: janusgraph-solr/src/test/resources/solr/core-template/schema.xml [hunk 1]
```diff
@@ -547,6 +547,7 @@
                prefixTree="geohash"/>
```
Comment (sjudeng) [inline janusgraph-solr/src/test/resources/solr/core-template/schema.xml]: Can you do me a favor and replace `org.locationtech.spatial4j.context.jts.JtsSpatialContextFactory` with `com.spatial4j.core.context.jts.JtsSpatialContextFactory` above? Otherwise I can open another PR with the fix. In testing this with Solr 5.5.4 (see [TESTING.md](https://github.com/JanusGraph/janusgraph/blob/master/TESTING.md#running-solr-tests-with-external-server)) I realized my change to the test schema in #400 broke backwards-compatibility. Incidentally the tests on this branch against Solr 5.5.4 all passed.

Comment (holleyism) [inline janusgraph-solr/src/test/resources/solr/core-template/schema.xml]: I can, but should it also be updated in the imports for JtsGeoshapeHelper.java and JtsGeoshapeHelperTest.java?

Comment (sjudeng) [inline janusgraph-solr/src/test/resources/solr/core-template/schema.xml]: No the client is still Solr 6, where correct package is `org.locationtech` ... it's just that the test schema.xml can be used when testing against other Solr server versions.

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 2]
```diff
@@ -332,6 +335,8 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
                 String collectionName = stores.getKey();
                 String keyIdField = getKeyFieldId(collectionName);
 
+                Map<String, String> fieldDeletes = new HashMap<String, String>(1) {{ put("set", null); }};
```
Comment (holleyism) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: Can't use ImmutableMap here since the value is null.  Will use Collections.unmodifiableMap here and ImmutableMap elsewhere.

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 3]
```diff
@@ -354,7 +359,17 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
                                     fieldDeletions.remove(indexEntry);
                                 }
                             }
-                            deleteIndividualFieldsFromIndex(collectionName, keyIdField, docId, fieldDeletions);
+                            SolrInputDocument doc = new SolrInputDocument();
+                            doc.addField(keyIdField, docId);
+                            for(IndexEntry v: fieldDeletions) {
+                                KeyInformation keyInformation = informations.get(collectionName, v.field);
+                                doc.setField(v.field, keyInformation.getCardinality() == Cardinality.SINGLE?
+                                    fieldDeletes:new HashMap<String, Object>(1) {{ put("remove", convertValue(v.value));}});
```
Comment (holleyism) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: Turns out I can't use ImmutableMap here either because when the solrj client parses the immutablemap, it doesn't like the geopoint.  I've added collections.unmodifiablemap in the appropriate places.
Caused by: org.apache.solr.client.solrj.impl.HttpSolrClient$RemoteSolrException: Error from server at http://127.0.0.1:62442/solr/vertex_shard1_replica1: Unable to parse shape given formats "lat,lon", "x y" or as WKT because java.text.ParseException: Unknown Shape definition [com.google.common.collect.ImmutableMap$Builder:com.google.common.collect.ImmutableMap$Builder@5e7022a2]

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 4]
```diff
@@ -332,6 +335,8 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
                 String collectionName = stores.getKey();
                 String keyIdField = getKeyFieldId(collectionName);
 
+                Map<String, String> fieldDeletes = Collections.unmodifiableMap(new HashMap<String, String>(1) {{ put("set", null); }});
```
Comment (sjudeng) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: Add `final`. Also I was suggesting `ImmutableMap.of(key,value)` to be more concise than `new HashMap...`, not because this should be immutable, so since ImmutableMap isn't possible can you revert back e.g. remove `Collections.unmodifiableMap` here and below? Sorry for the confusion, I should have been more clear in my comments.

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 5]
```diff
@@ -449,25 +464,29 @@ public void restore(Map<String, Map<String, List<IndexEntry>>> documents, KeyInf
...
-        StringBuilder sb = new StringBuilder();
-        for (IndexEntry fieldToDelete : fieldDeletions) {
-            doc.addField(fieldToDelete.field, fieldDeletes);
-            sb.append(fieldToDelete).append(",");
+    private Map<String,Object> buildFieldAdditions(List<IndexEntry> content, String collectionName, KeyInformation.IndexRetriever informations) throws BackendException {
+        Map<String,Object> docs = new HashMap<String,Object>();
+        for (IndexEntry addition: content) {
+            KeyInformation keyInformation = informations.get(collectionName, addition.field);
+            switch (keyInformation.getCardinality()) {
+                    case SINGLE:
```
Comment (sjudeng) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: Indent by 4 instead of 8 spaces here.

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java (hunk 4 above)
Comment (holleyism) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: got it.

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java (hunk 5 above)
Comment (holleyism) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: darn new eclipse setup.

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 6]
```diff
@@ -449,25 +461,29 @@ public void restore(Map<String, Map<String, List<IndexEntry>>> documents, KeyInf
...
+            }
         }
-
-        if (logger.isTraceEnabled())
-            logger.trace("Deleting individual fields [{}] for document {}", sb.toString(), docId);
-
-        UpdateRequest singleDocument = newUpdateRequest();
-        singleDocument.add(doc);
-        solrClient.request(singleDocument, collectionName);
+        return Collections.unmodifiableMap(docs);
```
Comment (sjudeng) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: You can revert this change as well ... unmodifiable not necessary since method is private and result is used internally.

**Comment by holleyism** [comment]
Squashed commits after review comments.

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 7]
```diff
@@ -354,7 +359,18 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
                                     fieldDeletions.remove(indexEntry);
                                 }
                             }
-                            deleteIndividualFieldsFromIndex(collectionName, keyIdField, docId, fieldDeletions);
```
Comment (jerryjch) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: Is it possible to keep it in the deleteIndividualFieldsFromIndex() private method? Like you have the buildFieldAdditions().

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 8]
```diff
@@ -449,25 +461,29 @@ public void restore(Map<String, Map<String, List<IndexEntry>>> documents, KeyInf
...
-        Map<String, String> fieldDeletes = new HashMap<String, String>(1) {{ put("set", null); }};
-
-        SolrInputDocument doc = new SolrInputDocument();
-        doc.addField(keyIdField, docId);
-        StringBuilder sb = new StringBuilder();
-        for (IndexEntry fieldToDelete : fieldDeletions) {
-            doc.addField(fieldToDelete.field, fieldDeletes);
-            sb.append(fieldToDelete).append(",");
+    private Map<String,Object> buildFieldAdditions(List<IndexEntry> content, String collectionName, KeyInformation.IndexRetriever informations) throws BackendException {
+        final Map<String,Object> docs = new HashMap<String,Object>();
```
Comment (jerryjch) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: final Map<String, Object> docs = new HashMap<>();

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 9]
```diff
@@ -368,14 +384,15 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
...
-
-                        for (IndexEntry e : mutation.getAdditions()) {
-                            final Object fieldValue = convertValue(e.value);
-                            doc.setField(e.field, isNewDoc
-                                    ? fieldValue : new HashMap<String, Object>(1) {{
-                                put("set", fieldValue);
-                            }});
-                        }
+                        // Build map of additions
+                        final Map<String,Object> adds = buildFieldAdditions(mutation.getAdditions(), collectionName, informations);
```
Comment (jerryjch) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: Map<String, Object>  -- space

File: janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java [hunk 10]
```diff
@@ -368,14 +384,15 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
...
+                        adds.keySet().stream().forEach(v-> {
+                            final KeyInformation keyInformation = informations.get(collectionName, v);
+                            final String key = keyInformation.getCardinality() == Cardinality.SINGLE ? "set" : "add";
```
Comment (jerryjch) [inline janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java]: Maybe a different name, like modifier?  'key' can be confused with index key, etc.

**Comment by sjudeng** [comment]
It looks like there are test failures in BerkeleySolrTest and ThriftSolrTest in the latest push (see https://travis-ci.org/JanusGraph/janusgraph/jobs/267438208).

**Comment by holleyism** [comment]
I'm not sure what happened.  It passes locally.  Trying again.

**Comment by holleyism** [comment]
The line numbers in the log don't appear to match the file.  Repushing the test file.

**Comment by holleyism** [comment]
The build doesn't appear to be picking up the correct file.  The first error stack is:
```
at org.junit.Assert.assertFalse(Assert.java:74)
	at org.janusgraph.graphdb.JanusGraphIndexTest.testCollection(JanusGraphIndexTest.java:1833)
	at org.janusgraph.graphdb.JanusGraphIndexTest.testIndexing(JanusGraphIndexTest.java:1727)
```
Line 1727 is supposed to be calling testCollection which is in the stack trace, but line 1833 (inside the testGeo method) is
```
assertCount(i + 1, tx.query().has(geoShapeProperty, Geo.INTERSECT, Geoshape.circle(0.0, 0.0, distance)).edges());
```

**Comment by holleyism** [comment]
Looks like issue 408 added a new check that was failing.  Merging and fixing.

**Comment by sjudeng** [comment]
Can you rebase on master and resolve the conflict in JanusGraphIndexTest?

**Comment by sjudeng** [comment]
Can you rebase instead of merge or squash to remove the merge commit?

**Comment by pluradj** [comment]
Fixes #327
